sync/sync_head.go

```diff
@@ -186,7 +193,68 @@ func (s *Syncer[H]) verify(ctx context.Context, newHead H) (bool, error) {
 			"reason", verErr.Reason)
 	}
 
-	return verErr.SoftFailure, err
+	return false, err
+}
+
+// verifySkipping will try to find such headers in range (subjHead, networkHeader)
+// that can be verified by subjHead, literally:
+//
+//	header.Verify(subjHead, candidate)
+//
+// and also such headers can verify `networkHeader`, literally
+//
+//	header.Verify(candidate, networkHeader)
+//
+// When such candidates cannot be found [NewValidatorSetCantBeTrustedError] will be returned.
+func (s *Syncer[H]) verifySkipping(ctx context.Context, subjHead, networkHeader H) error {
+	subjHeight := subjHead.Height()
+
+	diff := networkHeader.Height() - subjHeight
+	if diff <= 0 {
+		panic(fmt.Sprintf("implementation bug: diff is %d", diff))
+	}
+
+	for diff > 1 {
+		candidateHeight := subjHeight + diff/2
+
+		candidateHeader, err := s.getter.GetByHeight(ctx, candidateHeight)
+		if err != nil {
+			return err
+		}
+
+		if err := header.Verify(subjHead, candidateHeader); err != nil {
+			// candidate failed, go deeper in 1st half.
+			diff = diff / 2
+			continue
+		}
+
+		// candidate was validated properly, update subjHead.
+		subjHead = candidateHeader
+		// TODO: s.setSubjectiveHead(ctx, subjHead)
```

Comment:
Not sure should this happen here/right now/at all. 

AFAIR @Wondertan had ideas on this?

Reply:
I don't think we should set subjective head here.

Reply:
Reason?

Reply:
It's valid to set the subjective head here as verification succeeds against previously know subjective head.

Reply:
Uncommented, thanks.